### ISSUE TITLE
`save-place-find-file-hook` doesn't get added to `find-file-hook` from time to time

### ISSUE THREAD

**Issue by HyunggyuJang** [hlissner/doom-emacs#4759]
**What did you expect to happen?**
After `saveplace` got loaded, I expected `save-place-find-file-hook` to be added to `find-file-hook`

**What actually happened?**
To inspect how it behaves, I've inserted following code into my `config.el`
``` emacs-lisp
(use-package! saveplace
  :defer t
  :config
  (message "installed to find-file-hook? %s" (memq 'save-place-find-file-hook find-file-hook)))
```
And from the `*Messages*` buffer, I got
``` text
installed to find-file-hook? nil
```

**Additional details:**
As far as I debugged, it turns out that only `save-place-find-file-hook` doesn't get installed correctly -- other hooks handled by `save-place--setup-hooks` are intact and worked as expected.

**Steps to reproduce:**
1. Open Doom Emacs.
2. Launch Org agenda by hitting `Spc n a a`
3. Then my debug code spits 
   ``` text
   installed to find-file-hook? nil
   ```
I suspect that the `doom-first-file-hook` somewhat mangled `find-file-hook`. It only happens whenever I implicitly visit files, like with Org agenda command, and never occurs if I issue a command involving `find-file` to visit one.
**System information:**

https://pastebin.com/LwyCs0am

**Comment by HyunggyuJang** [comment]
Okay, I acknowledge that my debug code doesn't work no matter what file it first visits. However, my reproduction steps are still valid: You can inspect `find-file-hook`'s value after issuing `Spc n a a`. Then you should observe `save-place-find-file-hook` wasn't listed in that hook.

However, if you visit files with a command deliberately, then you can see the difference: there should be `save-place-find-file-hook` in `find-file-hook`'s value.

**Comment by hlissner** [comment]
The test is incorrect. `(memq 'save-place-find-file-hook find-file-hook)` only tests the buffer-local value of `find-file-hook` (which will likely be `(git-gutter t)` in most buffers. It will never return non-nil.

`save-place-find-file-hook` is added to its _global_ value, therefore `(memq 'save-place-find-file-hook (default-value 'find-file-hook))` would be better, but still not perfect, because `save-place-find-file-hook` isn't added to the hook under `after-find-file` or `doom-switch-buffer-hook` trigger, which will always occur _right_ after the package is loaded.

A better test would be:

```elisp
(defadvice! does-it-run? (&rest _)
  :before #'save-place-find-file-hook
  (message "It ran in %s!" (current-buffer)))
```

Now open a buffer and observe `It ran in ...` emitted in \*Messages\*, as intended.

As there is no issue here, I'll close this, but feel free to follow up here.

**Comment by HyunggyuJang** [comment]
Thanks for your detailed advice. 

First of all, I'd like to check whether I've understood what you pointed

Is it correct to assume that what you meant in
> but still not perfect, because `save-place-find-file-hook` isn't added to the hook until `after-find-file` or `doom-switch-buffer-hook` triggers, which do not occur immediately after the package is loaded (beyond the scope of an `after!` or `use-package!` + `:config` block), but shortly afterwards.
was `save-place-find-file-hook`, hence I guess the body of `save-place-mode`, getting evaluated only after `after-find-file` or `doom-switch-buffer-hook`?

I know that I'm picking your brain about this, but since I couldn't have the anatomy of how Doom Emacs down pat, I'm hoping you correct me if any misunderstanding.

From my digestion, I believe your `defadvice!` debug code should emit the message whenever I visit new file, which isn't yet loaded as buffer, but the thing is, after visiting Org agenda with a command of `SPC n a a`, your debug code gets paralyzed: 
Here comes my `org-agenda-files`
```elisp
("/Users/hynggyujang/org/inbox.org" "/Users/hynggyujang/org/gcal.org")
```
And this is what I get from your debug code:
```text
Doom loaded 215 packages across 35 modules in 2.773s [2 times]
HGJ: org configuration called
Press key for agenda command:
Loading /Users/hynggyujang/.emacs.d/.local/cache/recentf...done
It ran in inbox.org!
```
I believe it should print `It ran in gcal.org!` also, and furthermore, it doesn't spit anymore afterwards even if I visit a bunch of files not yet visited.

Thus, what I described in the reproductive steps is still valid. I couldn't narrow down what the causation was, but I hope you could direct me in this.

**Comment by HyunggyuJang** [comment]
Besides, after adding 
```elisp
(after! saveplace
  (add-hook 'find-file-hook 'save-place-find-file-hook t))
```
to my `config.el`, I could resolve the problem. However, it is an ad-hoc solution and I believe there should be some fundamental issues lurking in deeper sites.

**Comment by hlissner** [comment]
After some further investigation, I realized there was indeed a bug with `doom-first-file-hook`'s side effects being suppressed by an optimization hack in our org-agenda module. As of 9f08db8 it should be resolved. Sorry for the trouble, and thanks for bringing it to my attention!

**Comment by HyunggyuJang** [comment]
Thanks for your quick resolution! And I apologize my misbehavior that bringing your attention by opening duplicate issue. I'm not familiar with this kind of process, and I misunderstood what you meant in 'follow up here'  as open a new issue since I felt closed issue doesn't get processed. 

Anyway, I love your masterpiece, thanks.

**Comment by hlissner** [comment]
> Unfortunately, it causes yet another problem. One cannot use `org-agenda-clock-in` or any other org-agenda related commands that modifies one of the agenda files but not visiting them, thus not switching buffer to it.
> 
> By deep-delving related code. I found that it still makes some errors not related with org agenda as I indicated from the #4759, implicitly visited file buffer. To name one, `helpful--definition` also calls `find-file` command. I know that it is an extreme case to call describe commands as a first command to be issued; nonetheless, I realized that there are some leakages where `doom-first-file-hook` does not work as intended: It should be run for the first interactively opened file.
> 
> I found the command ground of those erroneous cases: Those binds `find-file-hook` with `let` to suppress unnecessary hook activations.
> 
> Thus, what I've figured out is that use `find-file-hook` as a canary of miners: If its value is `nil`, it might be within some non-interactive file visiting commands.
> 
> As a result, I'm using following workaround in the meantime while someone fixes this issues correctly:
> 
> ```emacs-lisp
> (defun doom-first-file-hook-h (&rest _)
>   (when (and after-init-time find-file-hook)
>     (run-hook-wrapped 'doom-first-file-hook #'doom-try-run-hook)
>     (set 'doom-first-file-hook nil)))
> ```
> 
> and for agenda buffers,
> 
> ```emacs-lisp
> (after! org
>   (defun +org--restart-mode-h ()
>     "Restart `org-mode', but only once."
>     (quiet! (org-mode-restart))
>     (delq! (current-buffer) org-agenda-new-buffers)
>     (remove-hook 'doom-switch-buffer-hook #'+org--restart-mode-h
>                  'local)))
> ```
> 
> just reverting your change since I found that in my use case, I never visit agenda files without routing agenda buffer if I meant to issue agenda command as my first command.

(Copied from [9f08db8](https://github.com/hlissner/doom-emacs/commit/9f08db8e334c1018c95fbb7046b9a3e20d3e2a28#commitcomment-48229267))

This should be fixed as of 0683861

**Comment by HyunggyuJang** [comment]
Thanks!! It works like a charm. I love the way you wrapped up things.

Can I ask one last thing? Although your solution did the great job for my initial concern, the things around `find-file-hook`, I see that with the change you made before for the `+org--restart-mode-h`, virtually forbids the commands actiaved in `org-agenda` buffer, e.g. clock in, `TODO` state change, etc.

I know that killing buffer was needed to suppress littered `recentf` list with `org-agenda-files`, but is there any workaround for it, anything you can come up with?

I may have to open yet another issue for it. Thanks for your kinder response.

Best regards,

&#x2013; Hyunggyu Jang
